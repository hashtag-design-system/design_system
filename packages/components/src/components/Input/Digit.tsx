import React from "react";
import { addClassnames } from "../../utils/styles";
import Input, { PasswordInputProps, ReactInputHTMLAttributes } from "./index";

// PasswordInputProps already has omitted props, omitted also in "this" props
export type Props = Omit<PasswordInputProps, "label" | "helpText" | "secondHelpText">;

export const Digit = React.forwardRef<HTMLInputElement, Props & Omit<ReactInputHTMLAttributes, "maxLength">>(
  ({ ...props }, ref) => {
    const { className, ...rest } = props;

    const classNames = addClassnames("input-digit shadow__form-4", props);

    return (
      <Input
        type="text"
        floatingPlaceholder={false}
        className={classNames}
        maxLength={1}
        characterLimit={false}
        width="auto"
        innerref={ref}
        {...rest}
      />
    );
  }
);

export default Digit;
