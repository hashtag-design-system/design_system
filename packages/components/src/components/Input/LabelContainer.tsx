import React from "react";
import { addClassnames } from "../../utils/styles";

type Props = {
  label?: string;
  withHelpText?: boolean;
  withIcon?: boolean;
  error?: boolean;
};

const LabelContainer: React.FC<Props & React.HTMLAttributes<HTMLElement>> = ({
  label,
  withHelpText = false,
  withIcon = false,
  error = false,
  children,
  ...props
}) => {
  const { className, ...rest } = props;
  let classNames = addClassnames("body-12", props);
  if (error) {
    classNames += " input__help-text error";
  } else {
    classNames += " input__help-text";
  }
  if (withIcon) {
    classNames += ` input__help-text__icon ${error ? "error" : ""}`;
  }

  // TODO: Make the input__help-text__icon classes in the stylesheets
  return (
    <div className="input__label-container" style={{ justifyContent: `${!label ? "flex-end" : ""}` }} {...rest}>
      {label}
      {withHelpText && <div className={`${classNames}`}>{children}</div>}
    </div>
  );
};

export default LabelContainer;
