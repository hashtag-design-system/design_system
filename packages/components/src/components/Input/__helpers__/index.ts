export * from "./Base/Base";
export * from "./BaseNumber";
export * from "./FieldContainer/FieldContainer";
export * from "./FixGroup/FixGroup";
export * from "./FloatingPlaceholder/FloatingPlaceholder";
export * from "./HelpTextContainer/HelpTextContainer";
export * from "./InputContainer/InputContainer";
export * from "./OuterFieldContainer/OuterFieldContainer";
export * from "./PasswordBoxes/PasswordBoxes";
