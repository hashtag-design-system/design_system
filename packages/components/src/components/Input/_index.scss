@import "./__helpers__/mixins";

.input {
  @include input-base;

  &__prefix {
    position: absolute;
    top: 18px;
    left: 14px;
  }

  &__wrapper {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    gap: 4px;
    width: 280px;

    &__base {
      display: flex;
      flex-direction: column;
      width: auto;
      position: relative;
      align-self: stretch;
    }
  }

  &__label-container {
    align-self: stretch;
    display: flex;
    flex-wrap: nowrap;
    justify-content: space-between;
    flex-direction: row;
    align-items: center;
    height: auto;
    font-family: "Roboto", sans-serif;
    font-weight: 500;
    font-style: normal;
    font-size: 14px;
    line-height: 150%;
    width: auto;
    color: $grey-9;
    margin-block: 0;
  }

  &__floating-label {
    position: absolute;
    top: 15px;
    left: 13px;
    width: auto;
    height: 18px;
    align-self: stretch;
    pointer-events: none;
    transition: all 0.2s ease-in-out;
    color: $grey-7;
    transform-origin: left top;
    outline: none;

    &:focus {
      @extend .active;
      color: $primary;
    }

    &:focus-within,
    &.active {
      @extend .input__floating-label;
      @include body;
      top: 8px;
      font-size: 10px;
      line-height: 100%;
      height: 10px;
    }
  }

  &__help-text {
    display: flex;
    flex-direction: row;
    align-items: flex-end;
    gap: 4px;
    align-self: flex-end;
    width: auto;
    height: 18px;
    margin-block: 0;
    color: $grey-7;

    &.error {
      color: $red-6;
    }
  }

  &-password {
    letter-spacing: 0.05em;
  }

  &__icon {
    position: absolute;
    width: 24px;
    height: 24px;
    right: 12px;
    top: calc(50% - 24px / 2);
  }

  @include input-placeholder-color;
}

.input,
.input-multiline {
  &.success {
    @include input-state-color($green-7);
  }

  &.error {
    @include input-state-color($red-6);
  }

  &.grey {
    &:focus {
      border-color: $grey-5;

      + .input__floating-label.active {
        color: $grey-7;
      }
    }
  }

  &:disabled,
  &.disabled,
  &[aria-disabled="true"] {
    @include disabled-state;
    background-color: $grey-4;
    border: none;
    color: $grey-8;
  }
}

.floating.input {
  &:focus-within {
    padding-top: 20px;
    padding-bottom: 9px;
  }
  &:not(:placeholder-shown) {
    padding-top: 20px;
    padding-bottom: 9px;
  }
}

.input-multiline {
  @extend .input__wrapper;
  // padding-top: set in the `element.style` attribute
  padding: 0px;
  height: auto;
  width: auto;
  gap: 0;

  .input {
    height: auto;
    border: none;
    align-self: stretch;
    padding: 0 12px 12px 12px !important;
    resize: vertical;
  }

  + .input__floating-label {
    top: 12px;

    &:focus-within,
    &.active {
      top: 8px;
    }
  }

  &:focus-within {
    @include input--focus;
  }
}

.input-digit {
  background-color: $grey-1;
  height: 80px;
  width: 70px;
  border-radius: 10px;
  border-color: $grey-1;
  font-size: 36px;
  line-height: 42px;
  text-align: center;
  justify-content: center;

  &:disabled,
  &.disabled,
  &[aria-disabled="true"] {
    box-shadow: none;
  }

  &-sequence {
    display: flex;
    flex-direction: column;
    align-items: center;
    flex: none;
    justify-content: center;
    width: auto;
    gap: 24px;

    &__wrapper {
      @extend .input-digit-sequence;
      @extend .input-digit-sequence__wrapper;
      flex-direction: row;
      gap: 20px;
    }

    .input__label-container {
      align-self: center;
    }
  }
}

.input-incr-dcr {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  text-align: center;
  margin: 8px;
  width: 120px;
  gap: 8px;
  height: 28px;

  .input {
    height: 28px;
    align-items: center;
    text-align: center;
    justify-content: center;
    padding: 0 !important;
    background-color: $grey-2;
    border-color: $grey-2;

    &:focus {
      background-color: $grey-3;
      border-color: $grey-3;
    }
  }

  &__icon {
    cursor: pointer;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 4px;
    padding: 0 !important;
    width: 28px;
    height: 28px;
    background-color: lighten($primary, 40);
    border-color: lighten($primary, 40);
    stroke: $primary;

    &:hover,
    &:focus-visible {
      background-color: lighten($primary, 30);
      border-color: lighten($primary, 30);
    }

    &:focus-visible {
      @include accessibility-3dp;
    }

    &.grey {
      background-color: $grey-4;
      border-color: $grey-4;
      cursor: default;
      stroke: $grey-7;
    }
  }

  &:disabled,
  &.disabled,
  &[aria-disabled="true"] {
    .input-incr-dcr__icon {
      @extend .grey;
    }
  }
}

.input-number {
  width: 120px;

  &__btn {
    background-color: $grey-2;
    position: absolute;
    overflow: hidden;
    top: 2px;
    right: 2px;
    height: 44px;
    border-radius: 0px 2px 2px 0px;
    border-left: 2px solid $grey-5;

    .btn {
      cursor: pointer;
      display: flex;
      justify-content: center;
      align-items: center;
      padding: 0 !important;
      width: 22px !important;
      height: 22px;
      border: none;
      box-shadow: none;
      stroke: $grey-6;

      &:hover {
        background-color: transparent;
        stroke: $primary;
      }

      &:nth-of-type(1) {
        border-radius: 0px 2px 0px 0px;
      }

      &:nth-of-type(2) {
        border-top: 2px solid $grey-5;
        border-radius: 0px 0px 2px 0px;
      }
    }
  }
}

.input-number,
.input-incr-dcr {
  .input {
    &::-webkit-outer-spin-button,
    &::-webkit-inner-spin-button {
      -webkit-appearance: none;
    }
  }
}
