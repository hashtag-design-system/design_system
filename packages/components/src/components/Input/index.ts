import Input from "./Input";

export type { Props as InputProps } from "./Input";
export type {
  GroupProps as InputNumberGroupProps,
  NumberProps as InputNumberProps,
  StepperProps as InputNumberStepperProps,
} from "./Number";
export type { Props as InputTextareaProps } from "./TextArea";

export default Input;
