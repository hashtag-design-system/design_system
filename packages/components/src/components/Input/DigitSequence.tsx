import React, { useEffect, useRef } from "react";
import keys from "../../config/keys";
import { InputHelpTextType } from "../../typings";
import { addClassnames } from "../../utils/styles";
import Input from "./Input";
import LabelContainer from "./__helpers__/LabelContainer";

export type Props = {
  digits: number;
  helpText?: InputHelpTextType;
  error?: boolean;
};

export const DigitSequence: React.FC<Props & React.HTMLAttributes<HTMLElement>> = ({
  digits = 4,
  helpText,
  error,
  ...props
}) => {
  const initialRef = useRef<HTMLInputElement>(null);
  const inputRefs = useRef<HTMLInputElement[] | null[]>([]);

  const { className, ...rest } = props;
  let classNames = addClassnames("input-digit-sequence", props);
  const state = error ? "error" : "default";

  useEffect(() => {
    if (initialRef.current) {
      initialRef.current.focus();
    }
  }, []);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>, i: number) => {
    if (e.target.value.length >= 1) {
      focus(i);
    }
  };

  const handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>, i: number) => {
    if (e.currentTarget.value.length >= 1) {
      focus(i);
    }
  };

  const focus = (i: number) => {
    if (inputRefs.current[i]) {
      inputRefs.current[i]?.focus();
    }
  };

  return (
    <div className={classNames} {...rest}>
      <div className="input-digit-sequence__wrapper" style={{ width: "auto" }}>
        {/* 
            Separated the components, so that no confusion is made in the browser 
            • 1st component should not have the same ref as the others, because it the first displayed & we set it by default to be focused on componentDidMount
            • All the other components, up to the LAST one, should have the same onChange and ref attributes
            • The last component should not have an onChange function, because it does not have a next component (ref) to focus on
          */}
        {[...Array(digits)].map((_, i) =>
          i === 0 ? (
            <Input.Digit
              ref={initialRef}
              onChange={e => handleChange(e, i)}
              onKeyPress={e => handleKeyPress(e, i)}
              state={state}
              key={`${keys.digitInput}${i}`}
            />
          ) : i === digits ? (
            <Input.Digit ref={ref => (inputRefs.current[i - 1] = ref)} state={state} key={`${keys.digitInput}${i}`} />
          ) : (
            <Input.Digit
              onChange={e => handleChange(e, i)}
              onKeyPress={e => handleKeyPress(e, i)}
              ref={ref => (inputRefs.current[i - 1] = ref)}
              state={state}
              key={`${keys.digitInput}${i}`}
            />
          )
        )}
      </div>
      {helpText && (
        <LabelContainer className="body-14" error={error ? true : false} withHelpText>
          {helpText.icon}
          {helpText.value}
        </LabelContainer>
      )}
    </div>
  );
};

export default DigitSequence;
