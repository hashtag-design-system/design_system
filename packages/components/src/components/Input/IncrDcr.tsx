import React, { useEffect, useReducer } from "react";
import { InputContextProvider, useClassnames, useDisabled } from "../../utils";
import Button, { ButtonFProps } from "../Button";
import { ComponentProps } from "../__helpers__";
import { InputNumberFProps } from "./index";
import { InputBaseState } from "./Input";
import { ACTIONS, BaseNumber, reducer, ReducerInitialStateType } from "./__helpers__";
import { AddIcon } from "./__icons__/AddIcon";
import { SubtractIcon } from "./__icons__/SubtractIcon";

// 11 states in total
const IncrDcrStates = [
  "focus-visible|increase",
  "focus-visible|decrease",
  "hover",
  "hover|increase",
  "hover|decrease",
  "active",
  "active|increase",
  "active|decrease",
] as const;
export type IncrDcrState = typeof IncrDcrStates[number] | InputBaseState;

export const initialState: ReducerInitialStateType = {
  min: 1,
  max: 9999,
  value: 1,
  defaultValue: 1,
  isDisabled: false,
  hasShiftKey: false,
};

export type Props = {
  btnProps?: ButtonFProps;
  onBtnClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
}

export type FProps = Props & Omit<
  InputNumberFProps,
  "state" | "helptext" | "secondhelptext" | "floatingplaceholder" | "label" | "optional" | "prefix" | "suffix"
> &
  Pick<ComponentProps<"input", false, IncrDcrState>, "state">;

const IncrDcr: React.FunctionComponent<FProps> = ({
  min = 1,
  max = 9999,
  step = 1,
  value: propsValue,
  defaultValue = min,
  width = "3rem",
  overrideOnChange,
  state = "default",
  btnProps,
  onValue,
  onBtnClick,
  ...props
}) => {
  const [classNames, rest] = useClassnames("input-incr-dcr", props);
  const isDisabled = useDisabled(props, state);
  const [{ value }, dispatch] = useReducer(reducer, initialState, (): typeof initialState => {
    return {
      ...initialState,
      value: propsValue || defaultValue || min,
      isDisabled,
      min,
      max,
    };
  });

  const handleOperation = (e: React.MouseEvent<HTMLButtonElement>, operation: "increment" | "decrement", stepNumber = step) => {
    e.preventDefault();
    dispatch({ type: ACTIONS[operation.toUpperCase()], payload: { step: stepNumber } });
    if (onBtnClick) {
      onBtnClick(e);
    }
  };

  useEffect(() => {
    if (onValue) {
      const parsedValue = parseFloat(value.toString() || min.toString());
      onValue(parsedValue);
    }
  }, [value]);

  const active = state === "active";
  const leftActive =
    (parseFloat(value.toString()) > min || state === "active|decrease" || state === "hover|decrease" || active) &&
    state !== "active|increase" &&
    !isDisabled;

  const rightActive =
    (parseFloat(value.toString()) < max || !value || state === "active|increase" || state === "hover|increase" || active) &&
    state !== "active|decrease" &&
    !isDisabled;

  return (
    <div className="input-incr-dcr__container">
      <Button
        disabled={!leftActive ? true : false}
        className={`input-incr-dcr__btn ${state === "hover|decrease" ? "hover" : ""} ${
          state === "focus-visible|decrease" ? "focus-visible" : ""
        }`}
        aria-label="Decrement"
        onClick={e => handleOperation(e, "decrement")}
        data-testid="input-incr-dcr-decrease"
        {...btnProps}
      >
        <SubtractIcon />
      </Button>
      <InputContextProvider
        value={{
          ...rest,
          value,
          dispatch,
          defaultValue,
          state,
          min,
          max,
          step,
          className: classNames,
          width,
          overrideOnChange,
        }}
      >
        <BaseNumber data-testid="input-incr-dcr" />
        <Button
          disabled={!rightActive ? true : false}
          className={`input-incr-dcr__btn ${state === "hover|increase" ? "hover" : ""} ${
            state === "focus-visible|increase" ? "focus-visible" : ""
          }`}
          aria-label="Increment"
          onClick={e => handleOperation(e, "increment")}
          data-testid="input-incr-dcr-increase"
          {...btnProps}
        >
          <AddIcon />
        </Button>
      </InputContextProvider>
    </div>
  );
};

IncrDcr.displayName = "InputIncrDcr";

export default IncrDcr;
