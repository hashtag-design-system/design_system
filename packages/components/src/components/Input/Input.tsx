import React from "react";
import { IconPropType } from "../../typings";
import { InputContextProvider } from "../../utils/contexts";
import { generateInputId } from "../../utils/hooks/useInputId";
import { InputAutocompleteType } from "../__helpers__";
import Digit from "./Digit";
import DigitSequence from "./DigitSequence";
import IncrDcr from "./IncrDcr";
import { InputBaseFProps } from "./index";
import Multiline from "./Multiline";
import Number from "./Number";
import Password from "./Password";
import { Base, FieldContainer, InputContainer } from "./__helpers__";

const InputBaseStates = ["default", "focus", "disabled"] as const;
export type InputBaseState = typeof InputBaseStates[number];

export const InputStates = [...InputBaseStates, "hover", "error", "success"] as const;
export type InputState = typeof InputStates[number];

export type InputHelpTextType = {
  value: string;
  icon?: IconPropType;
  error?: boolean;
  transparent?: boolean;
};

type State = {
  id: FProps["id"];
  value: FProps["value"];
};

export type Props = {
  defaultValue?: React.ReactText;
  helptext?: InputHelpTextType;
  secondhelptext?: InputHelpTextType;
  characterLimit?: boolean;
  floatingplaceholder?: boolean | { now: boolean };
  label?: string;
  allowClear?: boolean;
  autoComplete?: InputAutocompleteType;
  optional?: boolean;
  forwardref?: React.ComponentPropsWithRef<"input">["ref"];
  passwordboxes?: React.ReactNode;
};

export type FProps = Props & InputBaseFProps<InputState>;

export type SBProps = Omit<Props, "passwordboxes"> &
  Pick<FProps, "placeholder" | "state" | "allowClear" | "suffix" | "prefix" | "type" | "maxLength">;

export default class Input extends React.Component<FProps, State> {
  static Base: typeof Base;

  static Password: typeof Password;

  static Multiline: typeof Multiline;

  static Digit: typeof Digit;

  static DigitSequence: typeof DigitSequence;

  static Number: typeof Number;

  static IncrDcr: typeof IncrDcr;

  static displayName = "Input";

  state: State = {
    id: this.props.id || "",
    value: this.props.value || this.props.defaultValue || "",
  };

  componentDidMount = () => {
    if (!this.state.id) {
      this.setState(prevState => ({
        ...prevState,
        id: generateInputId(),
      }));
    }
  };

  handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    // `overrideOnChange` Prop is being handle by the <Base /> helper component
    const { onChange } = this.props;

    this.setState({ value: e.target.value });

    if (onChange) {
      onChange(e);
    }
  };

  render() {
    const { id, value } = this.state;

    const { value: propsValue, overrideOnChange } = this.props;
    // console.log(id);

    return (
      <InputContextProvider
        value={{
          ...this.props,
          id,
          value: overrideOnChange && propsValue !== undefined ? propsValue : value,
          onChange: overrideOnChange ? this.props.onChange : this.handleChange,
          overrideOnChange,
        }}
      >
        <InputContainer>
          <FieldContainer />
        </InputContainer>
      </InputContextProvider>
    );
  }
}

// 6 (+1 <Base /> helper) sub-components in total
// ! Always add sub-components after declaration inside the component
Input.Base = Base;
Input.Password = Password;
Input.Multiline = Multiline;
Input.Digit = Digit;
Input.DigitSequence = DigitSequence;
Input.Number = Number;
Input.IncrDcr = IncrDcr;
