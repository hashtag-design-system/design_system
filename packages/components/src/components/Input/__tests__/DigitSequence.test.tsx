import { render, screen } from "@testing-library/react";
import userEvent, { specialChars } from "@testing-library/user-event";
import Input from "../index";
import { OpenEye } from "../__icons__";

describe("<Input.DigitSequence />", () => {
  test("default behaviour", () => {
    render(<Input.DigitSequence numberOfDigits={4} focusOnRender />);
    const inputs = screen.queryAllByTestId("input-digit");

    expect(inputs).toHaveLength(4);
    expect(document.activeElement).toBe(inputs[0]);
    inputs.forEach(input => {
      expect(input).toHaveValue("");
      expect(input).not.toHaveClass("error");
    });
  });
  test("onChange basic functionality", () => {
    const onChange = jest.fn((val: string) => val);
    render(<Input.DigitSequence onChange={val => onChange(val)} numberOfDigits={4} />);
    const inputs = screen.queryAllByTestId("input-digit");

    inputs.forEach((input, i) => {
      if (i + 1 !== inputs.length) {
        userEvent.type(input, "A");
        expect(input).toHaveValue("A");
      }
    });

    const times = inputs.length - 1;
    const mockResults = onChange.mock.results;

    // Check onChange.mock.results
    expect(onChange).toHaveBeenCalledTimes(times + 1);
    expect(mockResults[mockResults.length - 1].value).toBe("AAA");
    expect(document.activeElement).toBe(inputs[times]);
    expect(inputs.map(input => input.getAttribute("value")).join("")).toBe("AAA");
  });
  test("with helptext(+ icon)", () => {
    render(<Input.DigitSequence helptext={{ value: "Test", icon: <OpenEye /> }} />);
    const helptext = screen.getByTestId("input-digit-sequence-help-text");

    expect(helptext).toBeVisible();
    expect(helptext).toHaveTextContent("Test");
    expect(helptext).toMatchSnapshot();
    expect(helptext.hasChildNodes()).toBe(true);
    expect(helptext.childNodes).toHaveLength(2);
    // Only <svg /> (icon) is measured as a childElement
    expect(helptext.childElementCount).toBe(1);
    expect(screen.getByTestId("icon")).toBeVisible();
  });
  test("with error={true}", () => {
    render(<Input.DigitSequence error />);

    screen.queryAllByTestId("input-digit").forEach(input => expect(input).toHaveClass("error"));
  });
  test("with error={true} & helptext", () => {
    render(<Input.DigitSequence error helptext={{ value: "Test", error: true }} />);

    screen.queryAllByTestId("input-digit").forEach(input => expect(input).toHaveClass("error"));
    expect(screen.getByTestId("input-digit-sequence-help-text")).toHaveClass("error");
  });
  test("paste value from clipboard", () => {
    render(<Input.DigitSequence numberOfDigits={4} />);
    const inputs = screen.queryAllByTestId("input-digit");

    userEvent.type(inputs[0], "0");
    userEvent.paste(inputs[1], "tes");

    expect(inputs.map(input => input.getAttribute("value")).join("")).toBe("0tes");
  });
  test("backspace to move to previous input, if currentTarget.value.length < 1", () => {
    const onChange = jest.fn(val => val);
    render(<Input.DigitSequence numberOfDigits={4} onChange={val => onChange(val)} />);
    const inputs = screen.queryAllByTestId("input-digit");

    userEvent.type(inputs[0], "0");
    userEvent.paste(inputs[1], "tes");

    expect(inputs.map(input => input.getAttribute("value")).join("")).toBe("0tes");

    userEvent.type(inputs[3], specialChars.backspace);
    userEvent.type(inputs[3], specialChars.backspace);

    expect(inputs[3]).toHaveValue("");
    expect(inputs[2]).toHaveValue("");
    expect(inputs.map(input => input.getAttribute("value")).join("")).toBe("0t");
  });
});
