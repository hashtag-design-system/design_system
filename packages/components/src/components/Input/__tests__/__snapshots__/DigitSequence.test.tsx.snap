// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input.DigitSequence /> with helptext(+ icon) 1`] = `
<div
  class="input__help-text__container body-12  second"
  data-testid="input-digit-sequence-help-text"
>
  Test
  <svg
    class="icon"
    data-testid="icon"
    fill="none"
    height="24"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M14.945 12c0 1.592-1.319 2.882-2.945 2.882-1.627 0-2.945-1.29-2.945-2.882S10.373 9.118 12 9.118c1.627 0 2.945 1.29 2.945 2.882z"
    />
    <path
      clip-rule="evenodd"
      d="M12 5C9.112 5 6.846 6.263 5.223 7.669a15.052 15.052 0 00-3.043 3.673c-.24.411-.24.905 0 1.316.408.7 1.422 2.27 3.043 3.673C6.846 17.737 9.113 19 12 19c2.888 0 5.154-1.264 6.777-2.669a15.05 15.05 0 003.043-3.673c.24-.411.24-.905 0-1.316a15.052 15.052 0 00-3.043-3.673C17.154 6.263 14.887 5 12 5zm-8.724 6.954a13.815 13.815 0 012.783-3.36C7.533 7.318 9.513 6.235 12 6.235c2.487 0 4.467 1.083 5.941 2.36a13.812 13.812 0 012.783 3.36.09.09 0 01.014.045.09.09 0 01-.014.046 13.813 13.813 0 01-2.783 3.36c-1.474 1.275-3.454 2.359-5.941 2.359-2.487 0-4.467-1.084-5.941-2.36a13.814 13.814 0 01-2.783-3.36.09.09 0 01-.014-.045.09.09 0 01.014-.046z"
      fill-rule="evenodd"
    />
  </svg>
</div>
`;
